src/anomalib/metrics/spro.py

```diff
@@ -0,0 +1,158 @@
+
+        Example:
+            To update the metric state for the current batch, use the ``update`` method:
+
+            >>> spro.update(preds, _, labels)
+        """
+        assert masks is not None
+        self.targets.append(masks)
+        self.preds.append(predictions)
+
+    def compute(self) -> torch.Tensor:
+        """Compute the macro average of the SPRO score across all masks in all batches.
+
+        Example:
+            To compute the metric based on the state accumulated from multiple batches, use the ``compute`` method:
+
+            >>> spro.compute()
+            tensor(0.5433)
+        """
+        targets = dim_zero_cat(self.targets)
+        preds = dim_zero_cat(self.preds)
+        return spro_score(preds, targets, threshold=self.threshold, saturation_config=self.saturation_config)
+
+
+def spro_score(
+    predictions: torch.Tensor,
+    targets: torch.Tensor,
+    threshold: float = 0.5,
+    saturation_config: dict | None = None,
+) -> torch.Tensor:
+    """Calculate the SPRO score for a batch of predictions.
+
+    Args:
+        predictions (torch.Tensor): Predicted anomaly masks
+        targets: (torch.Tensor): Ground truth anomaly masks with non-binary values and, original height and width
+        threshold (float): When predictions are passed as float, the threshold is used to binarize the predictions.
+        saturation_config (dict): Saturations configuration for each label (pixel value) as the keys.
+            Defaults: ``None`` (which the score is equivalent to PRO metric, but with the 'region' are
+            separated by mask files.
+
+    Returns:
+        torch.Tensor: Scalar value representing the average SPRO score for the input batch.
+    """
+    predictions = torch.nn.functional.interpolate(predictions.unsqueeze(1), targets.shape[1:])
+
+    # Apply threshold to binary predictions
+    if predictions.dtype == torch.float:
+        predictions = predictions > threshold
+
+    score = torch.tensor(0.0)
+    m = 0
+    # Iterate for each image in the batch
+    for i, target in enumerate(targets):
+        unique_labels = torch.unique(target)
+
+        # Iterate for each ground-truth mask per image
+        for label in unique_labels[1:]:
+            # Calculate true positive
+            target_per_label = target == label
+            true_pos = torch.sum(predictions[i] & target_per_label)
+
+            # Calculate the areas of the ground-truth
+            defect_areas = torch.sum(target_per_label)
+
+            if saturation_config is not None:
+                # Adjust saturation threshold based on configuration
+                saturation_per_label = saturation_config[label.int().item()]
+                saturation_threshold = torch.minimum(
+                    torch.tensor(saturation_per_label["saturation_threshold"]),
+                    defect_areas,
+                )
```

Comment:
I don't think this minimum is mentioned in the paper? I assume that if the saturation threshold is set correctly, it should always be smaller than defect area, or is this wrong?

Reply:
For this case, I was referring to the evaluation code I found here: https://github.com/nelson1425/EfficientAD?tab=readme-ov-file#mvtec-loco-dataset. It seems to be just a precautionary measure, in case the saturation_threshold is incorrectly set to a larger value than the defect area. However, providing a warning might be a better way to handle such a case instead of using this minimum.

![image](https://github.com/openvinotoolkit/anomalib/assets/5786636/114db65b-0690-41af-abac-a83b8ed4e024)

Reply:
It does seem like a precautionary measure to me as well and I agree that user should be warned in such case. Maybe @nelson1425 can provide more insight here.

Reply:
@jpcbertoldo I think Eq1 is this:
https://github.com/openvinotoolkit/anomalib/blob/2cdeb82eb79348a23325fb3d873862775366faa2/src/anomalib/metrics/spro.py#L164

But the one mentioned by @blaz-r is to avoid the saturation_threshold set to a larger value than the actual anomalous areas, which could cause wrong calculation. For example, if the actual anomalous area is only 100, but the threshold is set to 200, it means there is no way for the predictions to achieve a perfect score even if all anomaly areas are correctly predicted. `score=min(100/200, 1) = 0.5`

Reply:
I have added the warning. Along with the warning, I think we still need to change the threshold for the case when it is incorrectly set to a larger value than the defect area. Because having larger value in the config does not make sense and would cause incorrect score calculation.

https://github.com/openvinotoolkit/anomalib/blob/53aa07f9209dd19add239db8b147adf70ff880ed/src/anomalib/metrics/spro.py#L151-L158